Translate State and Lifecycle page to French (#12)
reactjs/fr.reactjs.org: content/docs/state-and-lifecycle.md

setInterval(tick, 1000);
```

[**Essayer sur CodePen**](http://codepen.io/gaearon/pen/gwoJZk?editors=0010)

> Dans cette section, nous allons apprendre à faire un composant `Horloge` vraiment réutilisable et isolé. Il mettra en place son propre minuteur et se mettra à jour tout seul chaque seconde.

Author: Au temps pour moi, j'avais mal compris les guidelines alors ! Cela dit je préfère ça comme ça, ça me fait bien tiquer aussi de voir des noms de variables et de composants en français 😅

> Nous commençons par isoler l'apparence de l'horloge :

Author: J'aurai bien aimé "encapsulé", mais ça fait très anglicisme selon moi. Happy to change it 😄

> Nous commençons par isoler l'apparence de l'horloge :

Reviewer: En fait « encapsulé » est le plus juste, et existe bien, mais il n'est pas assez connu de l'auditoire, je pense, à moins de viser déjà des devs chevronnés, ce qui n'est pas le but ici.

5. Supprimez la déclaration vide de la fonction.

```js
class Horloge extends React.Component {
  render() {
    return (
      <div>
        <h1>Bonjour, monde !</h1>
        <h2>Il est {this.props.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}
```

[**Essayer sur CodePen**](http://codepen.io/gaearon/pen/zKRGpo?editors=0010)

Le composant `Horloge` est maintenant défini comme une classe au lieu d'une fonction.

La méthode `render` sera appelée à chaque fois qu'une mise à jour a lieu, mais tant que l'on rend `<Horloge />` dans le même nœud DOM, une seule instance de la classe `Horloge` sera utilisée. Cela nous permet d'utiliser des fonctionalités supplémentaires telles que l'état local et les méthodes de cycle de vie.

## Ajouter Un État Local À Une Classe {#adding-local-state-to-a-class}

Nous allons déplacer la `date` des props à l'état en trois étapes :

1) Remplacez `this.props.date` avec `this.state.date` dans la méthode `render()` :

```js{6}
class Horloge extends React.Component {
  render() {
    return (
      <div>
        <h1>Bonjour, monde !</h1>
        <h2>Il est {this.state.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}
```

> 2) Ajoutez [un constructeur de classe](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Classes#Constructeur) qui assigne l'état initial `this.state`:

Reviewer: ```suggestion
2) Ajoutez [un constructeur de classe](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Classes#Constructeur) qui initialise `this.state`:
```

Aaaaah tu vois ! Là tu as pensé à mettre à jour le lien 👏 GG

En revanche on utilise beaucoup plus « affecte » que « assigne », un terme plus juridique en français. Mais ici, vu le contexte, ce sera plus fluide en disant juste « initialiser ».

Author: Oui j'ai essayé de le faire pour chaque lien, mais j'ai dû oublier celui où tu as commenté au-dessus.

  }
```

> Les composants à base de classe doivent toujours appeler le constructeur de base avec `props`.

Reviewer: ```suggestion
Les composants à base de classe devraient toujours appeler le constructeur de base avec `props`.
```

On va préserver le conditionnel. En fait ce n'est **pas** obligatoire, mais [ça évite des surprises](https://overreacted.io/fr/why-do-we-write-super-props/).

Et d'ailleurs, pour être blindés, on devrait en fait récupérer puis passer *tous* les arguments, car il y en a trois (props, contexte, réconcilieur), mais bon, ils n'en parlent pas et c'est très avancé, donc nous non plus 😉

Author: Noté merci 😄

3) Supprimez la prop `date` de l'élément `<Horloge />` :

```js{2}
ReactDOM.render(
  <Horloge />,
  document.getElementById('root')
);
```

Nous rajouterons plus tard le code du minuteur dans le composant lui-même.

> Le résultat est tel que suit :

Reviewer: ```suggestion
Le résultat ressemble à ceci :
```

Woaaah, « est tel que suit » ! On est dans les travées de l'Académie, là ! C'est très, très, très formel comme ton, on utilise beaucoup plus « ressemble à ceci » pour tous les « looks like this », ne serait-ce que par préservation du ton d'origine.

Author: En effet c'est plus léger 😅

  render() {
    return (
      <div>
        <h1>Bonjour, monde !</h1>
        <h2>Il est {this.state.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}

ReactDOM.render(
  <Horloge />,
  document.getElementById('root')
);
```

[**Essayer sur CodePen**](http://codepen.io/gaearon/pen/KgQpJd?editors=0010)

Ensuite, nous allons faire en sorte que le composant `Horloge` mette en place son propre minuteur et se mette à jour toutes les secondes.

> ## Ajouter Des Méthodes De Cycle De Vie À Une Classe {#adding-lifecycle-methods-to-a-class}

Reviewer: ```suggestion
## Ajouter des Méthodes de Cycle de Vie à une Classe {#adding-lifecycle-methods-to-a-class}
```

Allez, dernière fois que je corrige moi-même, s'il en reste je te fais confiance pour ajuster 😉

Dans des applications avec de nombreux composants, il est très important de libérer les ressources utilisées par les composants quand ils sont détruits.

> Nous voulons [mettre en place un minuteur](https://developer.mozilla.org/fr/docs/Web/API/WindowTimers/setInterval) quand une `Horloge` est rendue dans le DOM pour la première fois. On appelle cela "monter" avec React.

Reviewer: ```suggestion
Nous voulons [mettre en place un minuteur](https://developer.mozilla.org/fr/docs/Web/API/WindowTimers/setInterval) quand une `Horloge` apparaît dans le DOM pour la première fois. Le terme React pour cela est « montage ».
```

Plusieurs points :

- On ne dit presque jamais « rendre / rendu » en tant que verbe.  Ici, on parle en fait du montage, on peut donc décrire la situation plutôt.
- Guillemets français
- Puisqu'on a un gérondif plutôt qu'un infinitif au départ, et qu'on parle du processus plutôt que de l'action (ce n'est pas notre code qui monte le composant, il réagit au montage fait par le réconcilieur), on utilisera le « montage » prévu au glossaire.
- J'ai retravaillé la dernière phrase pour qu'elle coule plus naturellement.

> Nous voulons également [nettoyer le minuteur](https://developer.mozilla.org/fr/docs/Web/API/WindowTimers/clearInterval) quand le DOM produit par l'`Horloge` est supprimé. On appelle cela "démonter" avec React.

Reviewer: ```suggestion
Nous voulons également [nettoyer le minuteur](https://developer.mozilla.org/fr/docs/Web/API/WindowTimers/clearInterval) quand le DOM produit par l'`Horloge` est supprimé. En React, on parle de « démontage ».
```

Là aussi, phrase plus fluide sur la fin avec la transcription et « démontage » plutôt que « démonter ». On évite aussi d’être trop répétitif dans les tournures de phrase (oui, la VO l'était…), du coup j'ajuste un peu la forme.

  render() {
    return (
      <div>
        <h1>Bonjour, monde !</h1>
        <h2>Il est {this.state.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}
```

> Ces méthodes sont appelées "méthodes de cycle de vie".

Reviewer: ```suggestion
On les appelle des « méthodes de cycle de vie ».
```

Guillemets français, et on essaie de retirer la répétition de « méthodes ». Si la VO est fastidieuse, on n'est pas tenus d'en préserver la lourdeur 😁

> La méthode `componentDidMount()` est exécutée après que la sortie du composant ait été rendue sur le DOM. C'est un bon endroit pour mettre en place le minuteur :

Reviewer: ```suggestion
La méthode `componentDidMount()` est exécutée après que la sortie du composant a été injectée dans le DOM. C'est un bon endroit pour mettre en place le minuteur :
```

- Comme [le disent](https://github.com/reactjs/fr.reactjs.org/wiki/Meilleures-pratiques-de-traduction#après-que) les meilleures pratiques, « après que » est suivi de l'indicatif
- Tournure alternative pour éviter « rendre » comme verbe

  }
```

Notez que l'on a enregistré l'ID du minuteur directement sur `this`.

> Alors que `this.props` est mise en place par React lui-même et que `this.state` a un sens bien spécial, vous pouvez très bien ajouter manuellement d'autres champs sur la classe si vous avez besoin de stocker quelque chose qui ne participe pas au flux de données (comme un ID de minuteur).

Reviewer: ```suggestion
Alors que `this.props` est mis en place par React lui-même et que `this.state` a un sens bien spécial, vous pouvez très bien ajouter manuellement d'autres champs sur la classe si vous avez besoin de stocker quelque chose qui ne participe pas au flux de données (comme un ID de minuteur).
```

Micro-détail : puisqu'on parle de « champs », on va considérer `this.props` comme tel et accorder « mis » au masculin.

Author: D'accord, j'aurai plutôt dit que c'est une propriété de l'objet.

  render() {
    return (
      <div>
        <h1>Bonjour, monde !</h1>
        <h2>Il est {this.state.date.toLocaleTimeString()}.</h2>
      </div>
    );
  }
}

ReactDOM.render(
  <Horloge />,
  document.getElementById('root')
);
```

[**Essayer sur CodePen**](http://codepen.io/gaearon/pen/amqdNA?editors=0010)

Maintenant l'horloge se met à jour toutes les secondes.

> Récapitulons ce qu'il se passe et l'ordre dans lequel les méthodes sont invoquées :

Reviewer: ```suggestion
Récapitulons ce qui se passe et l'ordre dans lequel les méthodes sont invoquées :
```

« Ce qu’il » est rarement nécessaire en langage usuel, « ce qui » est plus naturel.

> 1) Quand `<Horloge />` est passé à `ReactDOM.render()`, React appelle le constructeur du composant `Horloge`. Puisque `Horloge` a besoin d'afficher l'heure actuelle, il initialise `this.state` avec un objet contenant l'heure actuelle. Nous mettrons cet état à jour plus tard.

Reviewer: ```suggestion
1) Quand `<Horloge />` est passé à `ReactDOM.render()`, React appelle le constructeur du composant `Horloge`. Puisque `Horloge` a besoin d'afficher l'heure actuelle, il initialise `this.state` avec un objet contenant l'heure actuelle. Nous mettrons cet état à jour par la suite.
```

« Par la suite » indique clairement qu’à ce stade de la page, on s'en est déjà occupé, là où « plus tard » a une ambiguïté sur le fait qu’on a déjà traité cette question ou non. C'est un **micro-détail**, hein…

> 2) React appelle ensuite la méthode `render()` du composant `Horloge`. C'est comme cela que React apprend ce qu'il faut afficher à l'écran. React met ensuite à jour le DOM pour correspondre à la sortir de la méthode `render()` du composant `Horloge`.

Reviewer: ```suggestion
2) React appelle ensuite la méthode `render()` du composant `Horloge`. C'est comme cela que React découvre ce qu'il faut afficher à l'écran. React met ensuite à jour le DOM pour correspondre à la sortir de la méthode `render()` du composant `Horloge`.
```

Le français « apprendre » est très connoté apprentissage / éducation / connaissance / savoir-faire, là où l'anglais l'utilise pour tout un tas de variations, notamment « découvrir » (en français ça existe dans « je viens de l'apprendre ! » ou « tu n’as pas appris la nouvelle ? »). Du coup « découvrir » est souvent plus approprié quand le sujet n'est ni sentient ni une IA.

3) Quand la sortie de l'`Horloge` est insérée dans le DOM, React appelle la méthode de cycle de vie `componentDidMount()`. À l'intérieur, le composant `Horloge` demande au navigateur de mettre en place un minuteur pour appeler la méthode `tick()` du composant une fois par seconde.

> 4) Chaque seconde, le navigateur appelle la méthode `tick()`. À l'intérieur, le composant `Horloge` planifie une mise à jour de l'interface utilisateur en appelant `setState()` avec un objet contenant l'heure actuelle. Grâce à l'appel à `setState()`, React sait que l'état a changé, et invoque à nouveau la méthode `render()` pour savoir ce qui devrait être affiché à l'écran. Cette fois, la valeur de `this.state.date` dans la méthode `render()` est différente, la sortie devrait donc inclure l'heure mise à jour. React met à jour le DOM en accord avec cela.

Reviewer: ```suggestion
4) À chaque seconde, le navigateur appelle la méthode `tick()`. À l'intérieur, le composant `Horloge` planifie une mise à jour de l'interface utilisateur en appelant `setState()` avec un objet contenant l'heure actuelle. Grâce à l'appel à `setState()`, React sait que l'état a changé, et invoque à nouveau la méthode `render()` pour savoir ce qui devrait être affiché à l'écran. Cette fois, la valeur de `this.state.date` dans la méthode `render()` est différente, la sortie va donc inclure l'heure mise à jour. React met à jour le DOM en fonction de ça.
```

- En français on utilise rarement « chaque » dans ce sens-là, en début de locution. On préfère « à chaque » (déjà corrigé plus haut) ou « tou(te)s les » (comme tu l'as déjà fait plus haut de ton propre chef).
- On a un indicatif, pas un conditionnel (que je trouverais un peu déconcertant sur ce coup !)
- « en accord avec cela » est un peu ampoulé (comme n‘importe que « cela » d'ailleurs), on peut fluidifier.
- Je ne sais pas ce qui s'est passé sur ton clavier, mais nombre de tes codes *inline* avaient un caractère chelou après le backtick ouvrant…

Author: Indeed, peut-être VSCode qui fait des siennes, étrange ! Je ferai attention pour la prochaine fois.

> 5) Si le composant `Horloge` est retiré du DOM, React appelle la méthode de cycle de vie `componentWillUnmount()` pour que le minuteur soit arrêté.

Reviewer: ```suggestion
5) Si le composant `Horloge` finit par être retiré du DOM, React appellera la méthode de cycle de vie `componentWillUnmount()` pour que le minuteur soit arrêté.
```

- Le *ever* indique une possibilité éventuelle, qu'on doit traduire de façon plus franche
- La concordance des temps en français nous demande un futur, du coup.

## Utiliser L'État Local Correctement {#using-state-correctly}

> Il y'a trois choses que vous devriez savoir à propose de `setState()`.

Reviewer: ```suggestion
Il y'a trois choses que vous devriez savoir à propos de `setState()`.
```

### Ne Modifiez Pas L'État Directement {#do-not-modify-state-directly}

Par exemple, ceci ne re-rendra pas un composant :

```js

> // Mauvais

Reviewer: ```suggestion
// Erroné
```

On n’est pas là pour émettre un jugement de valeur (« mauvais »), mais un truc factuel.  On a comme options « erroné » (un seul N) ou « incorrect », mais le premier est plus fort dans son rejet / avertissement.

> this.state.commentaire = 'Bonjour';

Reviewer: Attention à ne pas traduire les identifiants ; comme c'est le premier qui ne soit pas `Horloge`, au cas où tu aurais juste fait un rechercher-remplacer, je le signale (il en reste plein en-dessous, sois vigilant).

Author: Merci, indeed je n'étais pas sûr pour celui-là.

```

À la place, utilisez `setState()` :

```js
// Correct
this.setState({commentaire: 'Bonjour'});
```

> Le seul endroit où vous pouvez assigner `this.state` est dans le constructeur.

Reviewer: ```suggestion
Le seul endroit où vous pouvez affecter `this.state`, c’est le constructeur.
```

Comme `this.props` et `this.state` peuvent être mises à jour de façon asynchrone, vous ne devez pas vous baser sur leurs valeurs pour calculer le prochain état.

Par exemple, ce code peut échouer pour mettre à jour un compteur :

```js
// Mauvais
this.setState({
  compteur: this.state.compteur + this.props.increment,
});
```

Pour remédier à ce problème, utilisez la seconde forme de `setState()` qui accepte une fonction à la place d'un objet. Cette fonction recevra l'état précédent comme premier argument et les props au moment de la mise à jour comme second argument :

```js
// Correct
this.setState((state, props) => ({
  compteur: state.compteur + props.increment
}));
```

Nous avons utilisé une [fonction fléchée](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Fonctions/Fonctions_fl%C3%A9ch%C3%A9es) ci-dessus, mais une fonction normale fonctionne également :

```js
// Correct
this.setState(function(state, props) {
  return {
    compteur: state.compteur + props.increment
  };
});
```

### Les Mises À Jour De L'État Sont Fusionnées {#state-updates-are-merged}

Quand vous invoquez `setState()`, React fusionne les objets que vous donnez avec l'état actuel.

> Par exemple, votre état peut contenir plusieurs variables indépendentes :

Reviewer: ```suggestion
Par exemple, votre état peut contenir plusieurs variables indépendantes :
```

```js{4,5}
  constructor(props) {
    super(props);
    this.state = {

>       publications: [],

Reviewer: Indépendamment du fait qu'on ne traduit pas les identifiants, pour info, ”post” sera généralement traduit par « article », moins formel.

      commentaires: []
    };
  }
```

Ensuite, vous pouvez les mettre à jour indépendamment avec des appels séparés à `setState()` :

```js{4,10}
  componentDidMount() {
    recupererPublications().then(reponse => {
      this.setState({
        publications: reponse.publications
      });
    });

    recupererCommentaires().then(reponse => {
      this.setState({
        commentaires: reponse.commentaires
      });
    });
  }
```

La fusion n'est pas profonde, donc `this.setState({comments})` laisse `this.state.posts` intacte, mais remplace complètement `this.state.comments`.

## Les données descendent {#the-data-flows-down}

> Ni parent ni enfant ne peut savoir si un certain composant est à état ou non, et ne devraient pas se soucier de s'il est défini comme une fonction ou comme une classe.

Reviewer: ```suggestion
Ni parent ni enfant ne peuvent savoir si un certain composant est à état ou non, et ne devraient pas se soucier de savoir s'il est défini par une fonction ou une classe.
```

« Ni…ni » est une énumération qui constituent un sujet pluriel, donc on accorde.

> C'est pourquoi on dit souvent que l'état est local ou encapsulé. Il est impossible d'y accéder depuis un autre composant.

Reviewer: GG sur la simplification de la 2e phrase.

Author: Thanks 😄

Un composant peut choisir de passer son état à ses enfants via des props :

```js
<h2>Il est {this.state.date.toLocaleTimeString()}.</h2>
```

Cela marche également avec des composants définis par l'utilisateur :

```js
<DateFormatee date={this.state.date} />
```

> Le composant `DateFormatee` reçoit la `date` dans ses props et ne sait pas si elle vient de l'état de l'`Horloge`, des props de l'`Horloge`, ou était tapée à la main :

Reviewer: ```suggestion
Le composant `DateFormatee` reçoit la `date` dans ses props et ne sait pas si elle vient de l'état de l'`Horloge`, des props de l'`Horloge`, ou a été tapée à la main :
```

Concordance des temps.

```js
function DateFormatee(props) {
  return <h2>Il est {props.date.toLocaleTimeString()}.</h2>;
}
```

[**Essayer sur CodePen**](http://codepen.io/gaearon/pen/zKRqNB?editors=0010)

> On appelle souvent cela un flux de données "du haut vers le bas" ou "unidirectionnel". Un état local est toujours possédé par un composant spécifique, et toute donnée ou interface utilisateur dérivée de cet état ne peut affecter que les composants "en-dessous" de ce composant dans l'arbre de composants.

Reviewer: ```suggestion
On appelle souvent cela un flux de données « du haut vers le bas » ou « unidirectionnel ». Un état local est toujours possédé par un composant spécifique, et toute donnée ou interface utilisateur dérivée de cet état ne peut affecter que les composants « en-dessous » de celui-ci dans l'arbre de composants.
```

- Guillemets français
- Retrait de répétition

> Si vous imaginez un arbre de composants comme une cascade de props, chaque état de composant est une source d'eau supplémentaire qui rejoint la cascade à un point arbitraire, mais qui coule également vers le bas.

Reviewer: En français, « arbitraire » sous-entend souvent une décision humaine, je préfèrerais sans doute « quelconque », mais à toi de voir.

> Pour démontrer que tous les composants sont réellement isolés, nous pouvons créer un composant `Application` qui rend trois `<Horloge>`s :

Reviewer: ```suggestion
Pour démontrer que tous les composants sont réellement isolés, nous pouvons créer un composant `App` qui affiche trois `<Clock>`s :
```

Bon, du coup, j'ai dé-traduit les identifiants, mais c'est surtout le verbe « rendre » qu'il fallait transformer.
